central/imagev2/datastore/datastore_impl.go

```diff
@@ -0,0 +1,370 @@
+	"github.com/stackrox/rox/central/metrics"
+	"github.com/stackrox/rox/central/ranking"
+	riskDS "github.com/stackrox/rox/central/risk/datastore"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/errorhelpers"
+	"github.com/stackrox/rox/pkg/images/enricher"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/scancomponent"
+	"github.com/stackrox/rox/pkg/search"
+	pkgSearch "github.com/stackrox/rox/pkg/search"
+)
+
+var (
+	log = logging.LoggerForModule()
+
+	imagesSAC    = sac.ForResource(resources.Image)
+	allAccessCtx = sac.WithAllAccess(context.Background())
+
+	defaultSortOption = &v1.QuerySortOption{
+		Field: pkgSearch.LastUpdatedTime.String(),
+	}
+)
+
+type datastoreImpl struct {
+	keyedMutex *concurrency.KeyedMutex
+
+	storage store.Store
+
+	risks riskDS.DataStore
+
+	imageRanker          *ranking.Ranker
+	imageComponentRanker *ranking.Ranker
+}
+
+func newDatastoreImpl(storage store.Store, risks riskDS.DataStore,
+	imageRanker *ranking.Ranker, imageComponentRanker *ranking.Ranker) *datastoreImpl {
+	ds := &datastoreImpl{
+		storage: storage,
+
+		risks: risks,
+
+		imageRanker:          imageRanker,
+		imageComponentRanker: imageComponentRanker,
+
+		keyedMutex: concurrency.NewKeyedMutex(globaldb.DefaultDataStorePoolSize),
+	}
+	return ds
+}
+
+func (ds *datastoreImpl) Search(ctx context.Context, q *v1.Query) ([]pkgSearch.Result, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), "ImageV2", "Search")
+
+	return ds.storage.Search(ctx, q)
+}
+
+// Count returns the number of search results from the query
+func (ds *datastoreImpl) Count(ctx context.Context, q *v1.Query) (int, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), "ImageV2", "Count")
+
+	return ds.storage.Count(ctx, q)
+}
+
+func (ds *datastoreImpl) SearchImages(ctx context.Context, q *v1.Query) ([]*v1.SearchResult, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), "ImageV2", "SearchImages")
+
+	results, err := ds.Search(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+	var images []*storage.ImageV2
+	var existing []search.Result
+	for _, result := range results {
+		image, exists, err := ds.storage.GetImageMetadata(ctx, result.ID)
+		if err != nil {
+			return nil, err
+		}
+		// The result may not exist if the object was deleted after the search
+		if !exists {
+			continue
+		}
+		images = append(images, image)
+		existing = append(existing, result)
+	}
+
+	return convertMany(images, existing)
+}
+
+// SearchRawImages delegates to the underlying searcher.
+func (ds *datastoreImpl) SearchRawImages(ctx context.Context, q *v1.Query) ([]*storage.ImageV2, error) {
+	defer metrics.SetDatastoreFunctionDuration(time.Now(), "ImageV2", "SearchRawImages")
+
+	results, err := ds.Search(ctx, q)
```

Comment:
This is more a favor than a problem.  Can you add

// TODO(ROX-29943): Eliminate unnecessary 2 pass database queries